lib: lte_link_control: Update library documentation (#12268)
nrfconnect/sdk-nrf: doc/nrf/libraries/modem/lte_lc.rst

.. _lte_lc_readme:

LTE link control
################

.. contents::
   :local:
   :depth: 2

The LTE link control library provides functionality to control the LTE link on an nRF91 Series SiP.

The LTE link can be controlled through library configurations and API calls to enable a range of features such as specifying the Access Point Name (APN), switching between LTE network modes (NB-IoT or LTE-M), enabling GNSS support and power saving features such as Power Saving Mode (PSM) and enhanced Discontinuous Reception (eDRX).

The library also provides functionality that enables the application to receive notifications regarding LTE link parameters such as Radio Resource Control (RRC) connection state, cell information, and the provided PSM or eDRX timer values.

Configuration
*************

> To enable the library, set the option :kconfig:option:`CONFIG_LTE_LINK_CONTROL` option to ``y`` in the project configuration file :file:`prj.conf`.

Reviewer: ```suggestion
To enable the library, set the Kconfig option :kconfig:option:`CONFIG_LTE_LINK_CONTROL` to ``y`` in the project configuration file :file:`prj.conf`.
```

Establishing an LTE connection
==============================

> The following block of code shows how the API can be used to establish an LTE connection and get callbacks from the library:

Reviewer: ```suggestion
The following block of code shows how you can use the API to establish an LTE connection and get callbacks from the library:
```

                   k_sem_give(&lte_connected);
                   break;

           case LTE_LC_EVT_PSM_UPDATE:
           case LTE_LC_EVT_EDRX_UPDATE:
           case LTE_LC_EVT_RRC_UPDATE:
           case LTE_LC_EVT_CELL_UPDATE:
           case LTE_LC_EVT_LTE_MODE_UPDATE:
           case LTE_LC_EVT_TAU_PRE_WARNING:
           case LTE_LC_EVT_NEIGHBOR_CELL_MEAS:
           case LTE_LC_EVT_MODEM_SLEEP_EXIT_PRE_WARNING:
           case LTE_LC_EVT_MODEM_SLEEP_EXIT:
           case LTE_LC_EVT_MODEM_SLEEP_ENTER:
           case LTE_LC_EVT_MODEM_EVENT:
                   /* Handle LTE events */
                   break;

           default:
                   break;
           }
   }

   int main(void)
   {
           int err;

           printk("Connecting to LTE network. This may take a few minutes...\n");

           err = lte_lc_init_and_connect_async(lte_handler);
           if (err) {
                   printk("lte_lc_init_and_connect_async, error: %d\n", err);
                   return 0;
           }

           k_sem_take(&lte_connected, K_FOREVER);

           /* Continue execution... */
   }

> The code block demonstrates how the library can be utilized to asynchronously setup an LTE connection.

Reviewer: ```suggestion
The code block demonstrates how you can use the library to asynchronously set up an LTE connection.
```

Enabling power-saving features
==============================

> The recommended way of enabling power saving features is to use the library Kconfig options :kconfig:option:`CONFIG_LTE_PSM_REQ` and :kconfig:option:`CONFIG_LTE_EDRX_REQ`.

Reviewer: ```suggestion
The recommended way of enabling power saving features is to use the Kconfig options :kconfig:option:`CONFIG_LTE_PSM_REQ` and :kconfig:option:`CONFIG_LTE_EDRX_REQ`.
```

> PSM and eDRX can also be requested at run time using the :c:func:`lte_lc_psm_req` and :c:func:`lte_lc_edrx_req` function calls, however calling the functions during modem initialization can lead to conflicts with the value set by the Kconfig options.

Reviewer: ```suggestion
PSM and eDRX can also be requested at run time using the :c:func:`lte_lc_psm_req` and :c:func:`lte_lc_edrx_req` function calls.
However, calling the functions during modem initialization can lead to conflicts with the value set by the Kconfig options.
```

> The timer values requested by the modem can be configured with the following options:

Reviewer: ```suggestion
You can set the timer values requested by the modem using the following options:
```

   int main(void)
   {
           int err;

           printk("Connecting to LTE network. This may take a few minutes...\n");

           err = lte_lc_init_and_connect_async(lte_handler);
           if (err) {
                   printk("lte_lc_init_and_connect_async, error: %d\n", err);
                   return 0;
           }

           k_sem_take(&lte_connected, K_FOREVER);

           struct lte_lc_conn_eval_params params = {0};

           err = lte_lc_conn_eval_params_get(&params);
           if (err) {
                   printk("lte_lc_conn_eval_params_get, error: %d\n", err);
                   return 0;
           }

           /* Handle connection evaluation parameters... */

           /* Continue execution... */
   }

:c:struct:`lte_lc_conn_eval_params` lists all information that is available when performing connection pre-evaluation.

Modem sleep and TAU pre-warning notifications
=============================================

Modem firmware version 1.3.0 and higher supports receiving callbacks from the modem related to Tracking Area Updates (TAU) and modem sleep.
Based on these notifications, the application can alter its behavior to optimize for a given metric.

> For instance, TAU pre-warning notifications can be used to schedule data transfers prior to a TAU so that data transfer and TAU occurs within the same RRC connection window, thereby saving the potential overhead associated with the additional data exchange.

Reviewer: ```suggestion
For instance, TAU pre-warning notifications can be used to schedule data transfers before a TAU so that data transfer and TAU occurs within the same RRC connection window, thereby saving the potential overhead associated with the additional data exchange.
```

Modem sleep notifications can be used to schedule processing in the same operational window as the modem to limit the overall computation time of the nRF91 Series SiP.

To enable modem sleep and TAU pre-warning notifications, enable the following options:

* :kconfig:option:`CONFIG_LTE_LC_MODEM_SLEEP_NOTIFICATIONS`
* :kconfig:option:`CONFIG_LTE_LC_TAU_PRE_WARNING_NOTIFICATIONS`

> Additional configurations related to these features can be found in the API documentation for the library.

Reviewer: ```suggestion
For additional configurations related to these features, see the API documentation.
```
